src/utils/Cypher.ts

```diff
@@ -145,4 +139,111 @@ export default class Cypher {
       }
     }
   }
+
+  // This method will be reused in a Scheduler task that resumes migation
+  public static async handleCryptoSettingsChange(tenantID: string): Promise<void> {
+    const createDatabaseLock = LockingManager.createExclusiveLock(tenantID, LockEntity.DATABASE, 'migrate-settings-sensitive-data');
+    if (await LockingManager.acquire(createDatabaseLock)) {
+      try {
+        await Cypher.migrate(tenantID);
+        await Cypher.cleanupFormerSensitiveData(tenantID);
+        const keySettings = await SettingStorage.getCryptoSettings(tenantID);
+        keySettings.crypto.migrationToBeDone = false;
+        await SettingStorage.saveCryptoSettings(tenantID, keySettings);
+      } catch (err) {
+        throw new BackendError({
+          source: Constants.CENTRAL_SERVER,
+          module: MODULE_NAME,
+          method: 'handleCryptoSettingsChange',
+          message: `Sensitive Data migration for tenant with ID: ${tenantID} failed.`
+        });
+      } finally {
+        // Release the database Lock
+        await LockingManager.release(createDatabaseLock);
+      }
+    }
+  }
+
+  public static async migrate(tenantID: string): Promise<void> {
+    const cryptoSetting = await Cypher.getCryptoSetting(tenantID);
+    await Cypher.migrateSettings(tenantID, cryptoSetting);
+  }
+
+  public static async getSettingsWithSensitiveData(tenantID: string): Promise<SettingDB[]> {
+    // Get all settings per tenant
+    const settings = await SettingStorage.getSettings(tenantID, {},
+      Constants.DB_PARAMS_MAX_LIMIT);
+    // Filter settings with sensitiveData
+    return settings.result.filter((value: SettingDB) => {
+      if (value?.sensitiveData && !Utils.isEmptyArray(value?.sensitiveData)) {
+        return true;
+      }
+    });
+  }
+
+  public static async migrateSettings(tenantID: string, cryptoSetting: CryptoSetting): Promise<void> {
+
+    const settingsToMigrate = await Cypher.getSettingsWithSensitiveData(tenantID);
+    // If tenant has settings with sensitive data, migrate them
+    if (!Utils.isEmptyArray(settingsToMigrate)) {
+      // Migrate
+      for (const setting of settingsToMigrate) {
+        if (!setting.formerSensitiveData && Utils.isEmptyArray(setting.formerSensitiveData)) {
```

Comment:
I do not understand that test: if there's no formerSensitiveData (or an empty array of them), why launching the migration?

The sanity check is that at key change, if there's a previous former settings, they shall be replaced: current -> former, previous former -> deleted and new key is added.

Reply:
After a migration is finished the formerSensitiveData are cleaned up from settings. Which means that if the setting does not have formerSensitiveData (and migrationToBeDone is true), that setting has to be migrated - and the opposite if it has formerSensitiveData (and migrationToBeDone is true), that settings was migrated. After a migration is completed, the formerSensitiveData is cleaned and migrationToBeDone is set to false

Reply:
> After a migration is finished the formerSensitiveData are cleaned up from settings. Which means that if the setting does not have formerSensitiveData (and migrationToBeDone is true), that setting has to be migrated - and the opposite if it has formerSensitiveData (and migrationToBeDone is true), that settings was migrated. After a migration is completed, the formerSensitiveData is cleaned and migrationToBeDone is set to false

The migration code shall be reentrant, with reentrant code, you they only need to test for migrationToBeDone and then run again the code. Testing for the migration status is then not needed.

Reply:
I don't understand what should I change? Testing migrationToBeDone is needed to know if migration is finished or not. Testing for formerSensitiveData is needed to know if a particular setting what migrated or not.